l1-contracts/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -60,73 +77,256 @@ contract L1ERC20Bridge is IL1Bridge, IL1BridgeLegacy, ReentrancyGuard {
-        {
-            // Data to be used in delegate call to initialize the proxy
-            bytes memory proxyInitializationParams = abi.encodeCall(
-                IL2ERC20Bridge.initialize,
-                (address(this), l2TokenProxyBytecodeHash, _governor)
+    /// @dev We have to confirm that the deploy transactions succeeded.
+    /// @param _chainId of the chosen chain
+    /// @param _bridgeImplTxStatus The status of the L2 bridge implementation deploy transaction
+    /// @param _bridgeProxyTxStatus The status of the L2 bridge proxy deploy transaction
+    function finishInitializeChain(
+        uint256 _chainId,
+        ConfirmL2TxStatus calldata _bridgeImplTxStatus,
+        ConfirmL2TxStatus calldata _bridgeProxyTxStatus
+    ) external {
+        require(l2BridgeAddress[_chainId] == address(0), "L1EB B deployed 2");
+        require(bridgeProxyDeployOnL2TxHash[_chainId] != 0x00, "L1EB b. impl tx n sent");
+
+        /// if it already succeeded we can skip it
+        if (!bridgeImplTxSucceeded[_chainId]) {
+            require(
+                bridgehub.proveL1ToL2TransactionStatus(
+                    _chainId,
+                    bridgeImplDeployOnL2TxHash[_chainId],
+                    _bridgeImplTxStatus.batchNumber,
+                    _bridgeImplTxStatus.messageIndex,
+                    _bridgeImplTxStatus.numberInBatch,
+                    _bridgeImplTxStatus.merkleProof,
+                    TxStatus(uint8(_bridgeImplTxStatus.succeeded ? 1 : 0))
+                ),
+                "L1EB b. impl tx n conf" // not confirmed
             );
-            l2BridgeProxyConstructorData = abi.encode(bridgeImplementationAddr, _governor, proxyInitializationParams);
+            if (_bridgeImplTxStatus.succeeded) {
+                bridgeImplTxSucceeded[_chainId] = true;
+            }
         }
 
-        // Deploy L2 bridge proxy contract
-        l2Bridge = BridgeInitializationHelper.requestDeployTransaction(
-            zkSync,
-            _deployBridgeProxyFee,
-            l2BridgeProxyBytecodeHash,
-            l2BridgeProxyConstructorData,
-            // No factory deps are needed for the L2 bridge proxy, because it is already passed in previous step
-            new bytes[](0)
+        require(
+            bridgehub.proveL1ToL2TransactionStatus(
+                _chainId,
+                bridgeProxyDeployOnL2TxHash[_chainId],
+                _bridgeProxyTxStatus.batchNumber,
+                _bridgeProxyTxStatus.messageIndex,
+                _bridgeProxyTxStatus.numberInBatch,
+                _bridgeProxyTxStatus.merkleProof,
+                TxStatus(uint8(_bridgeProxyTxStatus.succeeded ? 1 : 0))
+            ),
+            "L1EB b. proxy tx n conf" // not confirmed
         );
+        delete bridgeImplDeployOnL2TxHash[_chainId];
+        delete bridgeProxyDeployOnL2TxHash[_chainId];
+        if ((_bridgeProxyTxStatus.succeeded) && bridgeImplTxSucceeded[_chainId]) {
```

Comment:
What if `_bridgeProxyTxStatus.succeeded == true` but `bridgeImplTxSucceeded[_chainId] == false?

Due to lines above:
- `bridgeImplDeployOnL2TxHash[_chainId] == 0`
- `bridgeProxyDeployOnL2TxHash[_chainId] == 0;`

As a result honest user would need to call `startErc20BridgeInitOnChain` again. Since deployment is done via `create2` then L1 -> L2 transaction for deployment proxy would fail always (!) and it would be impossible to finish bridge integration.

Reply:
BridgeProxyTxStatus = true means that there is a contract deployed at the l2Token beacon address, because the bridgeProxy calls the bridgeImpl at initialization.

Reply:
So it should be impossible to finish initialization

Reply:
`_bridgeProxyTxStatus == true`  only means the deployment succeeds. Now look into situation in L2, If there is no deployed contract at L2 bridge impl address, then there is no bytecode and any call/delegate call to it would succeed too. So when proxy deploys it will do empty initialization. What I am missing?

Reply:
Ah, ok, my mistake then. A call to an empty contract succeeds on L2 though? shouldn't it fail? But will revise

Reply:
> A call to an empty contract succeeds on L2 though?

Actually in both L1 and L2. Call to empty contract always succeeds

Reply:
This is very tricky though...

Reply:
Need to be documented very well

Reply:
Added comments.